% \VignetteIndexEntry{RnaSeqSampleSize: Sample size estimation by real data}
% \VignettePackage{RnaSeqSampleSize}
\documentclass[12pt]{article}
\textwidth 6.75in
\textheight 9.5in
\topmargin -.875in
\oddsidemargin -.06in
\evensidemargin -.06in

\usepackage{hyperref}
\hypersetup{
    colorlinks=true, %set true if you want colored links
    linktoc=all,     %set to all if you want both sections and subsections linked
    linkcolor=blue,  %choose some color if you want links to stand out
}

\usepackage{listings}
\usepackage{inconsolata}
<<echo=FALSE>>=
  options(width=60)

  listing <- function(x, options) {
    paste("\\begin{lstlisting}[breaklines=true]\n",
      x, "\\end{lstlisting}\n", sep = "")
  }
  knit_hooks$set(source=listing, output=listing)
@

\begin{document}
\title{RnaSeqSampleSize: Sample size estimation based on real RNA-seq data}
\author{Shilin Zhao}
\maketitle
\begin{abstract}
In this vignette, we demonstrated the application of RnaSeqSampleSize as an sample size estimation tool for RNA-seq data.
\\
\\
RnaSeqSampleSize package provides the following features:
\begin{itemize}
  \item Estimation of sample size or power by single read count and dispersion;
  \item Estimation of sample size or power by prior data;
  \item Power curve visualization;
  \item Optimization by power or sample size matrix;
\end{itemize}
\end{abstract}
\newpage
\tableofcontents
\newpage

\section{Introduction}
Sample size estimation is the most important issue in the design of RNA sequencing experiments.However, thousands of genes are quantified and tested for differential expression simul-taneously in RNA-seq experiments. The false discovery rate for statistic tests should be controlled. At the same time, the thousands of genes have widely distributed read counts and dispersions, which were often estimated by experience or set at the most conservative values in previous methods. As a result, the estimated sample size will be inaccurate or over-estimated.
\\
\\
To solve these issues, we developed a sample size estimation method based on the distributions of gene read counts and dispersions from real data. Datasets from the user's preliminary experiments or the Cancer Genome Atlas (TCGA) can be used as reference. The read counts and their related dispersions will be selected randomly from the reference based on their distributions, and from that, the power and sample size will be estimated and summarized.

\section{User friendly web interface}
A user friendly web interface for RnaSeqSampleSize package is provided at \url{http://cqs.mc.vanderbilt.edu/shiny/RnaSeqSampleSize/}. Most of the functions in Examples section can be performed in this website.

\section{Examples}
<<prepareData,echo=F,cache=F>>=
library(RnaSeqSampleSize)
@

\subsection{Estimation of sample size or power by single read count and dispersion}
\subsubsection{Power estimation}
For example, if we are estimating the power of finding significant genes for RNA-seq data with specified sample size, and we have the following parameters: 
\begin{itemize}
  \item Samples in each group: 63;
  \item Minimal fold change between two groups: 2;
  \item Minimal average read counts: 5;
  \item Maximal dispersion: 0.5;
  \item False discovery rate (FDR): 0.01;
\end{itemize}
As a result, the estimated power is 0.8 by \emph{est\_power} function.
<<singlePower,echo=TRUE,tidy=TRUE,cache=T>>=
example(est_power)
@
\subsubsection{Sample size estimation}
For example, if we are estimating the sample size for RNA-seq data to achieve desired power of finding significant genes, and we have the following parameters: 
\begin{itemize}
  \item Desired power of finding significant genes: 0.8;
  \item Minimal fold change between two groups: 2;
  \item Minimal average read counts: 5;
  \item Maximal dispersion: 0.5;
  \item False discovery rate (FDR): 0.01;
\end{itemize}
As a result, the estimated sample size is 63 by \emph{sample\_size} function.
<<singleSampleSize,each=TRUE,tidy=TRUE,cache=T>>=
example(sample_size)
@
\subsection{Estimation of sample size or power by prior data}
\subsubsection{Power estimation}
For example, if we are estimating the power of finding significant genes for RNA-seq data with specified sample size, and we have the following parameters:
\begin{itemize}
  \item Samples in each group: 65;
  \item Minimal fold change between two groups: 2;
  \item Prior data: TCGA READ data;
  \item False discovery rate (FDR): 0.01;
\end{itemize}
<<distributionPower,echo=TRUE,tidy=TRUE,cache=T>>=
example(est_power_distribtuion)
@
Here we demonstrated the power estimation by prior data in three different situations. 
\begin{itemize}
  \item If we are intesested in all genes, we can use repNumber parameter to specify random number of genes to perform power estimation;
  \item If we are only intesested in a list of genes, we can use selectedGenes parameter to specify the list of genes to perform power estimation; 
  \item If we are only intesested a specified pathway, we can use pathway and species parameters to specify the genes in a pathway to perform power estimation. 
\end{itemize}
\\
\\
As a result, we use \emph{est\_power\_distribtuion} function and find the estimated power is 0.81 for random genes, 0.77 for specified gene list, and 0.77 for genes in Glycolysis and Gluconeogenesis (pathway 00010) pathway.

\subsubsection{Sample size estimation}
For example, if we are estimating the power of finding significant genes for RNA-seq data with specified sample size, and we have the following parameters:
\begin{itemize}
  \item Desired power of finding significant genes: 0.8;
  \item Minimal fold change between two groups: 2;
  \item Prior data: TCGA READ data;
  \item False discovery rate (FDR): 0.01;
\end{itemize}
As a result, we use \emph{sample\_size\_distribution} function and find the estimated sample size is 65 for random genes.
<<distributionSampleSize,echo=TRUE,tidy=TRUE,cache=T>>=
example(sample_size_distribution)
@

\subsection{Power curve visualization}
For example, if we are going to compare the power of finding significant genes for different false discovery rate, and we have the following parameters:
\begin{itemize}
  \item Samples in each group: 63;
  \item Minimal fold change between two groups: 2;
  \item Minimal average read counts: 5;
  \item Maximal dispersion: 0.5;
  \item False discovery rate (FDR): 0.01 and 0.05;
\end{itemize}
<<singlePowerCurves,echo=TRUE,tidy=TRUE,cache=T>>=
example(est_power_curve)
@
As a result, the relation between power and sample size can be estimated by \emph{est\_power\_curve} function and the power curves can be generated by \emph{ plot\_power\_curve} function.

\subsection{Optimization by power or sample size matrix}
For example, if the budget is limited, we need to balance the number of replications and sequence depth. We can use the \emph{optimize\_parameter} function to find the relation between sample size, read counts, and estimated power. And then the optimized parameters can be determined.
<<optimazation,echo=TRUE,tidy=TRUE,cache=F>>=
example(optimize_parameter)
@
As a result, the estimated power distribution indicates that the number of replications plays a more significant role in determining the power than the number of read counts.
\end{document}
